src/libertem/io/dataset/hdf5.py

```diff
@@ -291,6 +364,159 @@ def _get_tiles_normal(self, tiling_scheme, tileshape_nd, dest_dtype="float32"):
+                tile_slice_with_offset_1 = Slice(origin=origin_1, shape=shape_1)
+                # sometimes a second slice is not necessary at the end
+                if origin_1[0] + 1 < ds_shape[0]:
+                    origin_2 = (origin_1[0] + 1,) + (0,) + (o[2], o[3])
+                    shape_2 = Shape(
+                        (s[0],) + (s[1] - ds_shape[1] + origin_1[1],) + (s[2], s[3]),
+                        sig_dims=sig_dims
+                    )
+                    tile_slice_with_offset_2 = Slice(origin=origin_2, shape=shape_2)
+                    dest_origin_2 = (0,) + (shape_1[1],) + (0, 0)
+                    dest_tile_slice_2 = Slice(origin=dest_origin_2, shape=shape_2)
+            dest_origin_1 = ([0] * len(o))
+            dest_tile_slice_1 = Slice(origin=dest_origin_1, shape=shape_1)
+        return (
+            tile_slice_with_offset_1, dest_tile_slice_1, tile_slice_with_offset_2, dest_tile_slice_2
+        )
+
+    def _get_tile_slice_with_negative_offset(
+        self, tile_slice, slices_with_offset, offset, ds_shape, sig_dims
+    ):
+        o, s = tile_slice.origin, tile_slice.shape
+        abs_offset = -1 * offset
+        tile_slice_with_offset_1, dest_tile_slice_1, tile_slice_with_offset_2, dest_tile_slice_2 = (
+            slices_with_offset
+        )
+        tile_end_pos = (
+            np.ravel_multi_index(o[:-sig_dims], ds_shape[:-sig_dims])
+            + np.ravel_multi_index(tuple(a - 1 for a in s[:-sig_dims]), ds_shape[:-sig_dims])
+        )
+        # don't yield tiles otherwise
+        if tile_end_pos >= abs_offset:
+            tile_start_pos = (
+                np.ravel_multi_index(o[:-sig_dims], ds_shape[:-sig_dims])
+            )
+            # at the origin
+            if tile_start_pos < abs_offset:
+                adjusted_offset = min(abs_offset, abs(tile_start_pos + offset))
+                shape_1 = Shape(
+                    (s[0],) + (s[1] - adjusted_offset,) + (s[2], s[3]), sig_dims=sig_dims
+                )
+                tile_slice_with_offset_1 = Slice(
+                    origin=tuple([0] * (len(o) - sig_dims)) + tuple(o[-sig_dims:]),
+                    shape=shape_1
+                )
+                dest_origin_1 = (0,) + (adjusted_offset,) + (0, 0)
+                dest_tile_slice_1 = Slice(origin=dest_origin_1, shape=shape_1)
+            else:
+                frame_pos_with_offset = (
+                    np.ravel_multi_index(o[:-sig_dims], ds_shape[:-sig_dims]) + offset
+                )
+                origin_1 = np.unravel_index(
+                    frame_pos_with_offset,
+                    ds_shape[:-sig_dims]
+                ) + o[-sig_dims:]
+                # single slice required for reading
+                if origin_1[1] + s[1] <= ds_shape[1]:
+                    shape_1 = Shape(s, sig_dims=sig_dims)
+                    tile_slice_with_offset_1 = Slice(origin=origin_1, shape=shape_1)
+                # two slices required for reading
```

Comment:
> it is not clear to me why you can just look at `ds_shape[1]` to make these decisions

Sorry! The offset code for HDF5 is hard coded for 4D right now, and I'm yet to generalize it.

> why there are two slices required

Because, in case of a 4D HDF5 dataset, a single slice object can't access frames from two different rows (necessary when an offset is applied) and slice objects don't support concatenation. So I've to slice twice and combine the results.
(Please correct me if I got something wrong, thanks!)

Reply:
> Sorry! The offset code for HDF5 is hard coded for 4D right now, and I'm yet to generalize it.

No worries, good to know!

> Because, in case of a 4D HDF5 dataset, a single slice object can't access frames from two different rows (necessary when an offset is applied) and slice objects don't support concatenation. So I've to slice twice and combine the results.

Sounds right, you could also split into two tiles in this case - the `DataSet` classes have quite a bit of freedom on how to return the data, in the sense that a single tile *can* have less frames in it than defined in the tiling scheme, for example for border conditions like this.

Reply:
> you could also split into two tiles in this case - the DataSet classes have quite a bit of freedom on how to return the data, in the sense that a single tile can have less frames in it than defined in the tiling scheme

Thank you! I wish I had thought of that much earlier! I spent too much time on HDF5's offset and thinking about how to generalize it for n-D data, thanks again! :)